Added permanent-redirect-code

Stono opened the pull request:
Hey, 
This PR addresses https://github.com/kubernetes/ingress-nginx/issues/2715 (I hope).  
It makes the permanent redirect code 308 by default (as this is more aligned to a permanent redirect) but also gives the ability to override it.

<< Golang noob... so check with caution

Please could you do me a build so I can test it out?

Stono commented:
/assign @aledbf

Stono commented:
Works a treat, without the code annotation:

```
> Accept: */*
>
< HTTP/1.1 308 Permanent Redirect
< Date: Tue, 24 Jul 2018 14:41:10 GMT
< Content-Type: text/html
< Content-Length: 180
< Connection: keep-alive
< Location: https://www.google.com
<
<html>
<head><title>308 Permanent Redirect</title></head>
<body bgcolor="white">
<center><h1>308 Permanent Redirect</h1></center>
<hr><center>nginx</center>
</body>
```

With the code annotation:
```
>
< HTTP/1.1 301 Moved Permanently
< Date: Tue, 24 Jul 2018 14:44:03 GMT
< Content-Type: text/html
< Content-Length: 178
< Connection: keep-alive
< Location: https://www.google.com
<
<html>
<head><title>301 Moved Permanently</title></head>
<body bgcolor="white">
<center><h1>301 Moved Permanently</h1></center>
<hr><center>nginx</center>
</body>
</html>
```

Could we get this merged?
Any change on a 0.18.0 ETA?

aledbf commented:
> Any change on a 0.18.0 ETA?

at least 2 weeks

antoineco commented on docs/user-guide/nginx-configuration/annotations.md:
```diff
@@ -43,6 +43,7 @@ You can add these Kubernetes annotations to specific Ingress objects to customiz
 |[nginx.ingress.kubernetes.io/limit-connections](#rate-limiting)|number|
 |[nginx.ingress.kubernetes.io/limit-rps](#rate-limiting)|number|
 |[nginx.ingress.kubernetes.io/permanent-redirect](#permanent-redirect)|string|
+|[nginx.ingress.kubernetes.io/permanent-redirect-code](#permanent-redirect-code)|number|
```
The type is not aligned with `GetStringAnnotation()` which you use later. But I think `number` is actually more suitable anyway (see comments below).

antoineco replied:
Maybe we could align the name on the ConfigMap field `http-redirect-code`?

antoineco replied:
Outdated annotation name.

antoineco commented on docs/user-guide/nginx-configuration/annotations.md:
```diff
@@ -366,6 +367,10 @@ To configure this setting globally for all Ingress rules, the `limit-rate-after`
 
+### Permanent Redirect Code
+
+This annotation allows you to modify the status code used for permanent redirects.  For example `nginx.ingress.kubernetes.io/permanent-redirect-code: 301` would return your redirect with a 301.
```
Annotations can't be int, you have to use single quotes around numeric values.

antoineco commented on internal/ingress/annotations/redirect/redirect.go:
```diff
@@ -73,14 +74,28 @@ func (a redirect) Parse(ing *extensions.Ingress) (interface{}, error) {
 		return nil, err
 	}
 
+	prc, err := parser.GetStringAnnotation("permanent-redirect-code", ing)
```
Maybe an int would be more suitable, no need to call `Atoi` then.

antoineco commented on internal/ingress/annotations/redirect/redirect.go:
```diff
@@ -73,14 +74,28 @@ func (a redirect) Parse(ing *extensions.Ingress) (interface{}, error) {
+	}
+
+	var prci = http.StatusPermanentRedirect
+	if prc != "" {
```
If you do use an `int`, the validation of the annotation could be something like
```golang
const defaultPermanentRedirectCode = http.StatusPermanentRedirect

if prc < http.StatusMultipleChoices || prc > http.StatusPermanentRedirect {
prc = defaultPermanentRedirectCode
}
```
(depending on what HTTP codes you believe are appropriate).

The const just makes the code mode readable, but more importantly you don't need 2 variables anymore.

Stono commented:
Hey @antoineco - I've pushed pushed your suggestions, however one thing that doesn't sit right with me is that i have renamed the annotation to `http-redirect-code`, as that makes sense, but really `permanent-redirect` should be renamed to `http-redirect` too, as having the type of redirect in that name infers the response code.

However, I didn't want to make such a breaking change.

antoineco commented on internal/ingress/annotations/redirect/redirect.go:
```diff
@@ -73,14 +73,25 @@ func (a redirect) Parse(ing *extensions.Ingress) (interface{}, error) {
 		return nil, err
 	}
 
+	const defaultPermanentRedirectCode = http.StatusPermanentRedirect
```
The default should remain http.StatusMovedPermanently for backward compatibility reasons.

antoineco replied:
I'd move this right above the definition of the Config struct, defaults should be clearly visible.

antoineco commented:
@Stono thanks for addressing my comments. You're completely right about the naming, we should not put the name of the http code inside the annotation name if the code is configurable.

I'd discuss the renaming of existing annotations in a different PR though, because, as you said, this change would be breaking backward compatibility unless we keep supporting the old and new annotations side by side  for a little while.

Stono commented:
@antoineco that's done, if you're happy i'd like to get this merged to ensure it's in the next release as it's super critical for me.

antoineco commented on internal/ingress/annotations/redirect/redirect.go:
```diff
@@ -73,14 +73,24 @@ func (a redirect) Parse(ing *extensions.Ingress) (interface{}, error) {
+		return nil, err
+	}
+
+	const defaultPermanentRedirectCode = http.StatusMovedPermanently
```
Please move the const declaration after the `import` block.

Stono commented:
Meh I give up mate, this back and forth is like a painful yoyo and unfortunately I don't have enough time to continue it.  Feel free to pick up this work otherwise i'll just build from my fork.

In hind sight I also don't agree with your renaming of `permanent-redirect-code` to `http-status-code`, as there is also `temporal-redirect`, which `https-status-code` wouldn't affect, so my changes here ARE only in relation to the code used for `permanent-redirects`.

Cheers

antoineco commented:
Sorry to read that. Both my comments were already mentioned before, you could have addressed them right away, in a single round trip.

All features need a test before we can merge them, I'm kindly referring you to the contribution guidelines.

Stono commented:
I was trying to help out in the five minutes spare time I have but as I made super clear I'm not a golang guy at all, you can see I moved the `const` definition, albeit to the wrong place - so it's not like I ignored your comments, I just simply didn't follow what you were asking.

I don't have the time to learn the testing frameworks for it right now, and I actually don't agree with your suggestions on the name of the `redirect-code`.  

So as I just explained, I'll pass on contributing to this project any further.  You're more than welcome to carry on the work as your redirect setup is incorrect and confusing to users.

antoineco commented:
You said you're not familiar with Go so I tried to help and share precise directions and good practices. A bit unfair to come and blame the reviewers afterwards for introducing confusion, isn't it?

I didn't make you use any name, I asked you about your opinion. Re-read my answer above.

Again, please check the contribution guidelines and show more empathy next time. We all try to help and deliver good software, and we all do it purely on our free time.

Stono commented:
Hey,
Thanks.  I'll add that later as I'm out at the moment.  Are we using "permanent-redirect-code" for now then?  It feels like it makes sense because of the undocumented temporal redirect and then things can be ratified in the other issue I raised.

antoineco commented:
This choice is up to you, both make sense to me according to what was discussed already.

Stono commented:
I've moved back to `permanent-redirect-code` and the unit tests pass, I've applied your patch and pushed that too.  However, I don't really understand the e2e tests, and don't follow your comment about http.ErrUseLastResponse either!  It also doesn't help that I don't know how to actually run them either.

I've just added you as a collaborator on my fork rather than messing about with patches.

Stono commented:
Thanks @antoineco

antoineco commented:
Found it. It was a simple misuse of the `gorequest` library. The function that determines the HTTP method (`Get(...)`) has to come first, otherwise the custom headers (`Set("Host", ...)`) and other parameters [get reset](https://github.com/parnurzeal/gorequest/blob/develop/gorequest.go#L201).

/lgtm
